### ISSUE TITLE
URL mangling with curly braces

### ISSUE THREAD

**Issue by ashemedai** [mastodon/mastodon#25467]
### Steps to reproduce the problem

1. I posted the valid URL `https://www.amsterdam.nl/veelgevraagd/?productid=%7BD5F9EF09-0F3A-4E59-8435-4873EB7CD609%7D` to someone.

### Expected behaviour

URL becomes clickable and remains intact

### Actual behaviour

The URL got mangled

### Detailed description

Resulting URL:
![image](https://github.com/mastodon/mastodon/assets/610865/5bef76f4-9281-4d49-8bcd-f2cc3651a31c)

Notice the cut off after the `productid=` and then the GUID between curly braces becomes standalone.

### Specifications

v4.1.2+nightly-20230614 (mastodon.social)

**Comment by Signez** [comment]
While investigating this issue, I discovered that the support of raw, non-encoded curly braces in URLs are not that obivous! Or at least, the characters `{`, `}`, `|`, `\`, `^`, `~`, `[`, `]`, and ``` ` ```, as they are considered _unsafe_ by [RFC 1738](https://www.rfc-editor.org/rfc/rfc1738#page-3).

Let's first note that if the URL is percent-encoded (like you posted in the issue in the first place), the links are shown properly on Mastodon's main.

![OK](https://github.com/mastodon/mastodon/assets/411336/ef4d7d6a-05db-41f4-b9f7-33b9bd67bc2b)

…but if I write the same URL but with curly braces non-encoded, indeed this happens:

![Not okay](https://github.com/mastodon/mastodon/assets/411336/3d2ffc14-e599-4224-ba7d-baba2b201e8d)

…and what about all the other "unsafe" characters? Well, surprise: some works, others don't. 

![Capture d’écran 2023-07-12 à 00 44 17](https://github.com/mastodon/mastodon/assets/411336/f830662e-453a-4bf0-b5aa-f11365d59643)

To sum up my findings:

* As soon as the URL is percent-encoded, there is no issue.
* If it's not:
  * `{`, `}`, `|`, `\`, `^`, `~`, `[`, `]`, and ``` ` ``` will not be considered part of a link if they are placed at the end of an URL.
  * `\`, `{`, `}` will never be part of a link and will cut it where they are placed.
